src/tools/gamemodelconverter/src/chunkfactory.cpp

```diff
@@ -0,0 +1,1837 @@
+{
+    char* rawData = ReadChunkRaw(cload);
+    if (strlen(rawData) == 0) { // case of null terminator
+        for (unsigned i = 0; i < data->data->chunkSize; i++) {
+            captainslog_dbgassert(rawData[i] == '\0', "Chunk %s size does not match data size", id);
+        }
+    } else {
+        captainslog_dbgassert(data->data->chunkSize == strlen(rawData) + 1, "Chunk %s size does not match data size", id);
+    }
+    AddData(data, id, "string", rawData, (void *)rawData);
+}
+
+template<typename> void Write_Chunk_String(ChunkSaveClass &csave, ChunkTreePtr &data, unsigned id)
+{
+    char* rawData = (char*)data->data->info->value;
+    if (strlen(rawData) == 0) { // case of null terminator
+        for (unsigned i = 0; i < data->data->chunkSize; i++) {
+            captainslog_dbgassert(rawData[i] == '\0', "Chunk %s size does not match data size", id);
+        }
+    } else {
+        captainslog_dbgassert(data->data->chunkSize == strlen(rawData) + 1, "Chunk %s size does not match data size", id);
+    }
+    WriteChunkRaw(csave, id, rawData, data->data->chunkSize);
+}
+#define READ_WRITE_CHUNK_STRING(id) \
+    void Read_##id(ChunkLoadClass &cload, ChunkTreePtr &data) { Read_Chunk_String<StringClass>(cload, data, #id); } \
+    void Write_##id(ChunkSaveClass &csave, ChunkTreePtr &data) { Write_Chunk_String<StringClass>(csave, data, id); } \
+
+// READ_WRITE_CHUNK
+template<typename T> void Read_Chunk_Struct(ChunkLoadClass &cload, ChunkTreePtr &data, const char *id)
+{
+    char* rawData = ReadChunkRaw(cload);
+    captainslog_dbgassert(data->data->chunkSize == sizeof(T), "Chunk %s size does not match data size", id);
+    AddData(data, id, typeid(T).name(), "", (void *)rawData);
+}
+
+template<typename T> void Write_Chunk_Struct(ChunkSaveClass &csave, ChunkTreePtr &data, unsigned id)
+{
+    char *rawData = (char *)data->data->info->value;
+    captainslog_dbgassert(data->data->chunkSize == sizeof(T), "Chunk %X size does not match data size", id);
+    WriteChunkRaw(csave, id, rawData, sizeof(T));
+}
+#define READ_WRITE_CHUNK(id, type) \
+    void Read_##id(ChunkLoadClass &cload, ChunkTreePtr &data) { Read_Chunk_Struct<type>(cload, data, #id); } \
+    void Write_##id(ChunkSaveClass &csave, ChunkTreePtr &data) { Write_Chunk_Struct<type>(csave, data, id); } \
+
+// READ_WRITE_CHUNK_ARRAY
+template<typename T> void Read_Chunk_Array(ChunkLoadClass &cload, ChunkTreePtr &data, const char *id)
+{
+    char *rawData = ReadChunkRaw(cload);
+    captainslog_dbgassert(data->data->chunkSize % sizeof(T) == 0, "Chunk %s size does not match data size", id);
+    std::vector<T> *arrayData = new std::vector<T>();
+    unsigned numElements = data->data->chunkSize / sizeof(T);
+
+    for (unsigned i = 0; i < numElements; i++) {
+        arrayData->push_back(*reinterpret_cast<T*>(rawData + i * sizeof(T)));
+    }
+
+    AddData(data, id, (std::string(typeid(T).name()) + "[]").c_str(), "", (void *)arrayData);
```

Comment:
This is wrong.

Passing a `std::vector<T> *` and ChunkArray will `delete[] (void*)`. It does not match.